### DISCUSSION TITLE
Extracting custom solidity errors

### DISCUSSION THREAD

**Discussion by isaac-martin** [wevm/wagmi#317] [Q&A]
Does anybody know the best way to extract custom solidity errors?
Taking the below solidity code (simplified)

```sol
    function claim(
        uint256 id
    ) public {

        if (!isPublicSaleActive) {
            revert PublicNotActive();
        }
       ....
    }
```

I want to be able to pull the error message in that you can get from etherscan - cant find a good way w/ this library (although this may be a downcoming in ethers...)

Wonder if it warrants a hook taht we could build to fetch this data easily..

<img width="882" alt="image" src="https://user-images.githubusercontent.com/8649366/161674588-26a0001c-2538-4fba-ac27-06de91068489.png">

**Comment by hyknerf** [comment]
Yes it's possible, use `useProvider()` hook.

```ts
const provider = useProvider();
```

we need some information from the transaction such as `from`, `to`, and `data` bytecode so we should get the transaction receipt first with `.getTransaction()`
```ts
provider.getTransaction(txHash)
```

and then use `.call()` on provider
```ts
provider.call({
    from: txResult.from,
    to: txResult.to,
    data: txResult.data,
  },
  txResult.blockNumber
);
```

the result will be EVM byte code like this:
```
0x08c379a000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000011696e76616c6964207369676e6174757265000000000000000000000000000000
```

but we only care about the last 64 character to read the error message and we want to convert it to ASCII you can do this
```ts
const lastArg = callResult.substring(
  callResult.length - 64,
  callResult.length - 1
);
let reason = "";
for (var n = 0; n < lastArg.length; n += 2) {
  console.log(lastArg.substring(n, n + 2));
  reason += String.fromCharCode(
    parseInt(lastArg.substring(n, n + 2), 16)
  );
}
// Gotcha!
console.log("Revert Reason:", reason);
```

Here is my working code
```ts
const provider = useProvider();

const [{ data: waitTxData }] = useWaitForTransaction({
  hash: "YOUR_TX_HASH",
});

useEffect(() => {
  if (!waitTxData?.transactionHash) return;

  provider.getTransaction(waitTxData?.transactionHash).then((response) => {
    console.log(response);
    provider
      .call(
        {
          from: response.from,
          to: response.to,
          data: response.data,
        },
        response.blockNumber
      )
      .then((callResult) => {
        const lastArg = callResult.substring(
          callResult.length - 64,
          callResult.length - 1
        );
        let reason = "";
        for (var n = 0; n < lastArg.length; n += 2) {
          reason += String.fromCharCode(
            parseInt(lastArg.substring(n, n + 2), 16)
          );
        }
        // Gotcha!
        console.log("Revert Reason:", reason);
      });
  });
}, [provider, waitTxData]);
```

**Comment by TheDojoSamurai** [comment]
This is very helpful. Is there a way to grab the revert message from `usePrepareContractWrite`?

**Reply by 0xCardiE** [reply]
Probably not as it seems just to fetch status, that is what ethers receipt provides, but would be really nice to have that info as well without another call to blockchain.
